src/tss2-tcti/tcti-cmd.c

```diff
@@ -0,0 +1,453 @@
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <inttypes.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <sys/prctl.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <sys/time.h>
+
+#include "tss2_tcti_cmd.h"
+
+#include "tcti-cmd.h"
+#include "tcti-common.h"
+#include "util/key-value-parse.h"
+#define LOGMODULE tcti
+#include "util/log.h"
+
+/*
+ * This function wraps the "up-cast" of the opaque TCTI context type to the
+ * type for the command TCTI context. If passed a NULL context the function
+ * returns a NULL ptr. The function doesn't check magic number anymore
+ * It should checked by the appropriate tcti_common_checks.
+ */
+TSS2_TCTI_COMMAND_CONTEXT*
+tcti_cmd_context_cast (TSS2_TCTI_CONTEXT *tcti_ctx)
+{
+    if (tcti_ctx == NULL)
+        return NULL;
+
+    return (TSS2_TCTI_COMMAND_CONTEXT*)tcti_ctx;
+}
+/*
+ * This function down-casts the cmd TCTI context to the common context
+ * defined in the tcti-common module.
+ */
+TSS2_TCTI_COMMON_CONTEXT*
+tcti_cmd_down_cast (TSS2_TCTI_COMMAND_CONTEXT *tcti_cmd)
+{
+    if (tcti_cmd == NULL) {
+        return NULL;
+    }
+    return &tcti_cmd->common;
+}
+
+TSS2_RC
+tcti_cmd_transmit (
+    TSS2_TCTI_CONTEXT *tcti_ctx,
+    size_t size,
+    const uint8_t *cmd_buf)
+{
+    TSS2_TCTI_COMMAND_CONTEXT *tcti_cmd = tcti_cmd_context_cast (tcti_ctx);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_cmd_down_cast (tcti_cmd);
+
+    TSS2_RC rc = tcti_common_transmit_checks (tcti_common, cmd_buf, TCTI_COMMAND_MAGIC);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    size_t bytes = fwrite (cmd_buf, 1, size, tcti_cmd->sink);
+    if (bytes != size) {
+    	LOG_ERROR ("Transmitting to subprocess failed: %s",
+    			strerror(errno));
+        return rc;
+    }
+
+    fflush(tcti_cmd->sink);
+
+    tcti_common->state = TCTI_STATE_RECEIVE;
+
+    return rc;
+}
+
+TSS2_RC
+tcti_cmd_cancel (
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TSS2_TCTI_COMMAND_CONTEXT *tcti_cmd = tcti_cmd_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_cmd_down_cast (tcti_cmd);
+    TSS2_RC rc;
+
+    rc = tcti_common_cancel_checks (tcti_common, TCTI_COMMAND_MAGIC);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    tcti_common->state = TCTI_STATE_TRANSMIT;
+    tcti_cmd->cancel = 1;
+
+    return rc;
+}
+
+TSS2_RC
+tcti_cmd_set_locality (
+    TSS2_TCTI_CONTEXT *tctiContext,
+    uint8_t locality)
+{
+    TSS2_TCTI_COMMAND_CONTEXT *tcti_cmd = tcti_cmd_context_cast (tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_cmd_down_cast (tcti_cmd);
+    TSS2_RC rc;
+
+    rc = tcti_common_set_locality_checks (tcti_common, TCTI_COMMAND_MAGIC);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+
+    tcti_common->locality = locality;
+    return TSS2_RC_SUCCESS;
+}
+// TODO do we need a define for this, this is coupled to device TCTI?
+#define TCTI_ASYNC
```

Comment:
Don't need that. The TCTI_ASYNC was only to handle kernel versions, where the driver didn't support poll()